### PR TITLE
CRAFTERCMS-1929; CRAFTERCMS-1930

### PR DESCRIPTION
**Pull request by dejan-brkic**
Rest APIs to add and remove remote.
Rest APIs to push to remote and pull from remote

### Ticket reference or full description of what's in the PR
https://github.com/craftercms/craftercms/issues/1929
https://github.com/craftercms/craftercms/issues/1930

### CODE REVIEW

File: src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java [hunk 1]
```diff
@@ -18,9 +18,7 @@
 
 package org.craftercms.studio.impl.v1.repository.git;
 
-import java.io.File;
-import java.io.IOException;
-import java.io.InputStream;
+import java.io.*;
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java]: Can we change that to the items we need?

File: src/main/resources/crafter/studio/database/createDDL.sql [hunk 2]
```diff
@@ -231,6 +231,26 @@ CREATE TABLE IF NOT EXISTS gitlog
   DEFAULT CHARSET = utf8
   ROW_FORMAT = DYNAMIC ;
 
+CREATE TABLE IF NOT EXISTS site_remote
+(
+  `id`                    BIGINT(20)    NOT NULL AUTO_INCREMENT,
+  `site_id`               VARCHAR(50)   NOT NULL,
+  `remote_name`           VARCHAR(255)   NOT NULL,
+  `remote_url`            VARCHAR(255)   NOT NULL,
+  `authentication_type`   VARCHAR(255)   NOT NULL,
```
Comment (sumerjabri) [inline src/main/resources/crafter/studio/database/createDDL.sql]: 255? too long?

File: src/main/resources/crafter/studio/database/createDDL.sql [hunk 3]
```diff
@@ -231,6 +231,26 @@ CREATE TABLE IF NOT EXISTS gitlog
...
+  `remote_username`       VARCHAR(255)   NULL,
+  `remote_password`       VARCHAR(255)   NULL,
+  `remote_token`          VARCHAR(255)   NULL,
+  `remote_private_key`    TEXT           NULL,
+  `salt`                  VARCHAR(255)   NULL,
```
Comment (sumerjabri) [inline src/main/resources/crafter/studio/database/createDDL.sql]: isn't this 16chars per code? maybe 20 but not 255, no?

File: src/main/resources/crafter/studio/database/upgrade-3.0.0.sql [hunk 4]
```diff
@@ -40,4 +40,23 @@ CREATE TABLE IF NOT EXISTS gitlog
 
 ALTER TABLE `site` ADD COLUMN `last_verified_gitlog_commit_id` VARCHAR(50) NULL ;
 
-INSERT INTO _meta (version) VALUES ('3.0.2.1') ;
+CREATE TABLE IF NOT EXISTS site_remote
```
Comment (sumerjabri) [inline src/main/resources/crafter/studio/database/upgrade-3.0.0.sql]: same notes as in the DDL file.

File: src/main/resources/crafter/studio/database/upgrade-3.0.1.sql [hunk 5]
```diff
@@ -14,4 +14,23 @@ CREATE TABLE IF NOT EXISTS gitlog
 
 ALTER TABLE `site` ADD COLUMN `last_verified_gitlog_commit_id` VARCHAR(50) NULL ;
 
-UPDATE _meta SET version = '3.0.2.1' ;
\ No newline at end of file
+CREATE TABLE IF NOT EXISTS site_remote
```
Comment (sumerjabri) [inline src/main/resources/crafter/studio/database/upgrade-3.0.1.sql]: same notes as in the DDL file

File: src/main/resources/crafter/studio/database/upgrade-3.0.2.1.sql [hunk 6]
```diff
@@ -0,0 +1,20 @@
+CREATE TABLE IF NOT EXISTS site_remote
```
Comment (sumerjabri) [inline src/main/resources/crafter/studio/database/upgrade-3.0.2.1.sql]: same notes as in the DDL file

File: src/main/resources/crafter/studio/database/upgrade-3.0.2.sql [hunk 7]
```diff
@@ -4,4 +4,23 @@ ALTER TABLE `gitlog` DROP COLUMN `commit_date` ;
 
 ALTER TABLE `site` ADD COLUMN `last_verified_gitlog_commit_id` VARCHAR(50) NULL ;
 
-UPDATE _meta SET version = '3.0.2.1' ;
\ No newline at end of file
+CREATE TABLE IF NOT EXISTS site_remote
```
Comment (sumerjabri) [inline src/main/resources/crafter/studio/database/upgrade-3.0.2.sql]: same notes as in the DDL file

File: src/main/webapp/default-site/scripts/rest/api/1/repo/add-remote.post.groovy [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+            paramsList.add("remote_url")
+        }
+    } catch (Exception exc) {
+        invalidParams = true
+        paramsList.add("remote_url")
+    }
+
+    // authentication_type
+    try {
+        if (StringUtils.isEmpty(authenticationType)) {
```
Comment (sumerjabri) [inline src/main/webapp/default-site/scripts/rest/api/1/repo/add-remote.post.groovy:81]: Where are we checking to see if the values map to the 4 things we support? (None, password, token, key)

**Review by sumerjabri — CHANGES_REQUESTED** [review]

File: src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java [hunk 9]
```diff
@@ -1550,11 +1550,231 @@ public boolean createSitePushToRemote(String siteId, String remoteName, String r
...
+            logger.debug("Inserting remote " + remoteName + " for site " + siteId + " into database.");
+            Map<String, String> params = new HashMap<String, String>();
+            params.put("siteId", siteId);
+            params.put("remoteName", remoteName);
+            params.put("remoteUrl", remoteUrl);
+            params.put("authenticationType", authenticationType);
+            params.put("remoteUsername", remoteUsername);
+
+            logger.debug("Generate salt to encrypt password");
+            String salt = RandomStringUtils.randomAlphanumeric(16);
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java]: I see the salt here, how does it relate, if at all, to the salt in the yaml config?

File: src/main/java/org/craftercms/studio/api/v1/repository/ContentRepository.java [hunk 10]
```diff
@@ -368,14 +365,44 @@
...
+     * @param remoteUrl remote url
+     * @param authenticationType authentication type
      * @param remoteUsername remote username
      * @param remotePassword remote password
-     * @return true if success
+     * @param remoteToken remote token
+     * @param remotePrivateKey remote private key
+     * @return true if operation was successful
+     */
+    boolean addRemote(String siteId, String remoteName, String remoteUrl, String authenticationType, String remoteUsername, String remotePassword, String remoteToken, String remotePrivateKey) throws InvalidRemoteUrlException, ServiceException;
```
[the remark this answers is not in the document]
Comment (dejan-brkic) [reply src/main/java/org/craftercms/studio/api/v1/repository/ContentRepository.java]: We don't have any service responsible for repo management. Right now only content service and site service execute repo operations.

File: src/main/java/org/craftercms/studio/impl/v1/dal/DataSourceInitializerImpl.java [hunk 11]
```diff
@@ -43,7 +43,8 @@
 public class DataSourceInitializerImpl implements DataSourceInitializer, DisposableBean {
 
     private final static Logger logger = LoggerFactory.getLogger(DataSourceInitializerImpl.class);
-    private final static String CURRENT_DB_VERSION = "3.0.2.1";
```
[the remark this answers is not in the document]
Comment (dejan-brkic) [reply src/main/java/org/craftercms/studio/impl/v1/dal/DataSourceInitializerImpl.java]: We have new database table. DB version must change. Not very happy with labeling mechanism, but it is functional.

File: src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java (hunk 1 above)
Comment (dejan-brkic) [reply src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java]: IDE changes to '*' after adding 4 classes from the package (i think 4 is the setting). I will change  the code and try to disable IDE setting

File: src/main/resources/crafter/studio/database/createDDL.sql (hunk 2 above)
Comment (dejan-brkic) [reply src/main/resources/crafter/studio/database/createDDL.sql]: Maybe. If your concern is space, I don't expect this table to grow too much

File: src/main/webapp/default-site/scripts/rest/api/1/repo/add-remote.post.groovy (hunk 8 above)
Comment (dejan-brkic) [reply src/main/webapp/default-site/scripts/rest/api/1/repo/add-remote.post.groovy:81]: When adding remote, there is no actual access to the remote. So we don't do any mapping, just store in DB. When executing pull/push, depending on authentication type stored, we map authentication provider.

File: src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java (hunk 9 above)
Comment (dejan-brkic) [reply src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java]: random salt from runtime (stored in db) + salt from yaml config is used to encrypt password

**Comment by dejan-brkic** [comment]
Added changes per review
Added API to list remotes

File: src/main/java/org/craftercms/studio/api/v1/repository/ContentRepository.java [hunk 12]
```diff
@@ -22,10 +22,7 @@
 import org.craftercms.studio.api.v1.dal.GitLog;
 import org.craftercms.studio.api.v1.exception.ContentNotFoundException;
 import org.craftercms.studio.api.v1.exception.ServiceException;
-import org.craftercms.studio.api.v1.exception.repository.InvalidRemoteRepositoryCredentialsException;
-import org.craftercms.studio.api.v1.exception.repository.InvalidRemoteRepositoryException;
-import org.craftercms.studio.api.v1.exception.repository.RemoteRepositoryNotBareException;
-import org.craftercms.studio.api.v1.exception.repository.RemoteRepositoryNotFoundException;
+import org.craftercms.studio.api.v1.exception.repository.*;
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/api/v1/repository/ContentRepository.java]: Please change to import what we need.

File: src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java [hunk 13]
```diff
@@ -1550,11 +1601,248 @@ public boolean createSitePushToRemote(String siteId, String remoteName, String r
...
+                        public void configure(Transport transport) {
+                            SshTransport sshTransport = (SshTransport)transport;
+                            sshTransport.setSshSessionFactory(getSshSessionFactory(remoteRepository, tempKey));
+                        }
+                    });
+                    pushCommand.call();
+                    Files.delete(tempKey);
+                    break;
+                default:
+                    throw new ServiceException("Unsupported autehentication type " + remoteRepository.getAuthenticationType());
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java]: Misspelling of "authentication" (it's not autehentication)

File: src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java [hunk 14]
```diff
@@ -1550,11 +1601,248 @@ public boolean createSitePushToRemote(String siteId, String remoteName, String r
...
+                        public void configure(Transport transport) {
+                            SshTransport sshTransport = (SshTransport)transport;
+                            sshTransport.setSshSessionFactory(getSshSessionFactory(remoteRepository, tempKey));
+                        }
+                    });
+                    pullCommand.call();
+                    Files.delete(tempKey);
+                    break;
+                default:
+                    throw new ServiceException("Unsupported autehentication type " + remoteRepository.getAuthenticationType());
```
Comment (sumerjabri) [inline src/main/java/org/craftercms/studio/impl/v1/repository/git/GitContentRepository.java]: Incorrect spelling of authentication.

**Review by sumerjabri — CHANGES_REQUESTED** [review]

**Pull request merged**